sdk/azcore/internal/exported/response_error.go

```diff
@@ -163,5 +169,38 @@ func (e *ResponseError) Error() string {
 	}
 	fmt.Fprintln(msg, separator)
 
-	return msg.String()
+	e.errMsg = msg.String()
+	return e.errMsg
+}
+
+// internal type used for marshaling/unmarshaling
+type responseError struct {
+	ErrorCode    string `json:"errorCode"`
+	StatusCode   int    `json:"statusCode"`
+	ErrorMessage string `json:"errorMessage"`
+}
+
+func (e ResponseError) MarshalJSON() ([]byte, error) {
+	// populate errMsg as needed
+	if e.errMsg == "" {
+		_ = e.Error()
+	}
+
+	return json.Marshal(responseError{
+		ErrorCode:    e.ErrorCode,
+		StatusCode:   e.StatusCode,
+		ErrorMessage: e.errMsg,
```

Comment:
I like `Error()` computing the message only once but just for the sake of clarity in my mind, do you really need to cache it on `ResponseError`? Couldn't you do
```suggestion
		ErrorMessage: e.Error(),
```

Reply:
The caching lets us preserve the output of `Error()` after the type has been unmarshaled.

Reply:
Your are correct though that this can be simplified, I've made that change.